internal/daemon/controller/testing.go

```diff
@@ -523,13 +523,15 @@ func NewTestController(t testing.TB, opts *TestControllerOpts) *TestController {
 		shutdownDoneCh: make(chan struct{}),
 		shutdownOnce:   new(sync.Once),
 	}
-	t.Cleanup(tc.Shutdown)
+	t.Cleanup(func() {
+		go tc.Shutdown()
```

Comment:
I believe there are a couple of tests that focus on validating the Shutdown function does not leak resources. For those specific tests, I believe that we should not utilize a cleanup function or a go routine when calling the Shutdown method. If those unit tests pass, I think we should feel comfortable with shutting down a worker/controller with go routines for unit tests that do not necessarily care about what the Shutdown method is doing.

Since the Shutdown method doesn't return an error, I feel like there is nothing to check against.

Reply:
I'm not convinced about this, sure, most of the time this probably works fine because the test process is still running while the cleanup is happening, but I really think we should at the very least ensure `TestMain` does not exist until all controllers have shut down successfully. I think I'd prefer if this PR did not run all these functions in goroutines, and we make any such changes in a separate PR. I could see us having something like a global waitgroup that we could add to and that `TestMain` could wait for before returning. Even that I'm not sure about, since `tc.Shutdown` can call `t.Error`, which is not allowed after the test has returned. Does that sound okay?

Reply:
I have an idea, but I'm not sure if it's worth the amount of complexity added for the benefit of improving how fast the tests run.

We could use a `TestMain(m *testing.M){}` to ensure that all the go routines in the cleanup functions finish running once the unit tests are done.

Example:

```
func TestMain(t *testing.M) {
  // setup a global variable for a sync.WaitGroup
  t.Run() // this will run the unit tests in the package
  // now we can call Wait() on the global waitgroup.
}
```

But at this point, we probably won't really get much of a performance improvement. Also this could cause more issues with tests timing out if people forget to call Done() on the waitgroup.

Reply:
Yeah that was my thinking too, but I also think it risks us missing test failures when there are errors in shutdown. So for this PR I'd prefer if we don't do the shutdowns in goroutines and ensure they all conclude before the test that spawned them. We can consider making shutdown changes in another PR. Does that make sense?

Reply:
Yea that makes sense to me, I will start reverting the changes to exclude the go routine calls for Shutdown.